tests/unit/core/mender-artifact_test.cpp

```diff
@@ -0,0 +1,320 @@
+/**
+ * @file      mender-artifact_test.cpp
+ * @brief     Unit Tests for mender-artifact.c
+ *
+ * Copyright Northern.tech AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include <sys/stat.h>
+#include <stdlib.h>
+#include <filesystem>
+#include <fstream>
+
+#include "mender-artifact.h"
+#include "mender-deployment-data.h"
+#include "mender-update-module.h"
+
+using namespace std;
+namespace fs = filesystem;
+
+#ifndef __has_feature
+// GCC does not have __has_feature
+#define __has_feature(feature) 0
+#endif
+
+#if ((__has_feature(address_sanitizer) || defined(__SANITIZE_ADDRESS__)) || (__has_feature(thread_sanitizer) || defined(__SANITIZE_THREAD__)))
+extern "C" {
+/* https://github.com/google/sanitizers/wiki/AddressSanitizerFlags#run-time-flags
+* Since we want to test cases where realloc will return null because of a
+* too large size, we need to set allocator_may_return_null=1 */
+const char *
+__asan_default_options() {
+    return "allocator_may_return_null=1";
+}
+const char *
+__tsan_default_options() {
+    return "allocator_may_return_null=1";
+}
+}
+#endif
+
+class MenderArtifactTest : public ::testing::Test {
+protected:
+    vector<uint8_t> artifact_data;
+    void           *data;
+    fs::path        artifact_path;
+    fs::path        script_file_name;
+
+    vector<uint8_t> CreateArtifact(string custom_script = "") {
+        string script;
+        if (custom_script.empty()) {
+            script = R"(#! /bin/sh
+            DIRNAME=$(dirname $0)
+            echo foobar > ${DIRNAME}/testdata
+
+            mender-artifact write module-image --compression none \
+            --type unit-test -n unit-test-artifact --device-type unit-test \
+            -f ${DIRNAME}/testdata -o ${DIRNAME}/unit-test-artifact.mender || exit 1)";
+        } else {
+            script = custom_script;
+        }
+
+        fs::path tmp_dir = fs::temp_directory_path();
+        script_file_name = tmp_dir / "test-script.sh";
+
+        ofstream os(script_file_name.c_str());
+
+        os << script;
+        os.close();
+
+        chmod(script_file_name.c_str(), S_IRUSR | S_IWUSR | S_IXUSR);
+        system(script_file_name.c_str());
+
+        artifact_path = tmp_dir / "unit-test-artifact.mender";
+        ifstream file(artifact_path);
+        return vector<uint8_t>((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
```

Comment:
Why `char`?

Reply:
As far as I can tell because the iterator reads chars from the ifstream `file`